### PR TITLE
Refactor ValidateLogin

### PR DESCRIPTION
**Pull request by robertjchen**
Thanks for the review! We're refactoring the old `ValidateEmail` call to a new `ValidateLogin` call that no longer requires chaining behavior. This flow is encountered when an account adds a secondary email and clicks on the resulting email validation link, leading them to verify their account on NewDot.

Due to the fact that we no longer chain the subsequent navigation on the result of the API call, we encountered a race condition where we were attempting to navigate before the `NavigationComponent` was full rendered. The solution was to temporarily cache the intended destination and do the redirect after the component had been fully loaded.

The corresponding backend API change has gone out already, and can be found here: https://github.com/Expensify/Web-Expensify/pull/34365

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/218822

#### Four use cases can occur with this flow:
1. A user is already logged into NewDot, and the secondary login is NOT validated
2. A user is already logged into NewDot, and their secondary login is already validated and/or the link is invalid
3. A user is NOT logged into NewDot, and their secondary login is NOT validated
4. A user is NOT logged into NewDot, and their secondary login is already validated and/or the link is invalid

#### Resulting Conditions
- In the first two cases, the user is simply redirected to their last open report/conversation with no error message displayed. Additionally, in Case 1, the account is silently validated behind the scenes.

- In the latter two cases, the user is redirected to the login page. In Case 4, an error message would displayed below the password field stating: "This set password link is invalid or has expired...", while in Case 3, the user would encounter a password prompt.

To test the cases described above:
1. Log into OldDot and add a Secondary Login from Account Settings e.g. mypersonal+validateTest@gmail.com
3. Log into NewDot, or not, depending on the test cases above.
4. Formulate the URL to emulate clicking a verification link `http://localhost:8080/v/<accountID>/<validationCode>` on local dev (use `script/clitools.sh generator:validateLink`), `http://staging.new.expensify.com/v/<accountID>/<validationCode>` on staging, and `http://staging.new.expensify.com/v/<accountID>/<validationCode>` on production (or, use an email you have access to)
5. Verify that the resulting conditions described hold true for all 4 cases

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

#### Web
Screen Shot 2022-08-23 at 6 18 48 PM

Screen Shot 2022-08-23 at 6 45 22 PM

#### Mobile Web
Screen Shot 2022-08-23 at 7 25 37 PM

#### Desktop
Not applicable, validation link opens browser

#### iOS
Not applicable, validation link opens browser

#### Android
Not applicable, validation link opens browser

### CODE REVIEW

**Comment by robertjchen** [comment]
Not ready for review just yet!

**Comment by robertjchen** [comment]
Still in progress, just need to fill out the tests section and provide screenshots once I get the App test environment working locally.

File: src/libs/Navigation/Navigation.js [hunk 1]
```diff
@@ -119,6 +121,9 @@ function isDrawerRoute(route) {
  */
 function navigate(route = ROUTES.HOME) {
     if (!canNavigate('navigate', {route})) {
+        // Store intended route if the navigator is not yet available,
+        // we will try again after the NavigationContainer is ready
+        pendingRoute = route;
```
Comment (roryabraham) [inline src/libs/Navigation/Navigation.js:126]: I'd recommend adding a log line here

**Comment by robertjchen** [comment]
Should be ready, thanks for the reviews! :shipit:

File: src/Expensify.js [hunk 2]
```diff
@@ -202,7 +206,7 @@ class Expensify extends PureComponent {
                 )}
 
                 <NavigationRoot
-                    onReady={this.setNavigationReady}
+                    onReady={this.setNavigationReady()}
```
Comment (roryabraham) [inline src/Expensify.js:209]: Shouldn't immediately execute this function

```suggestion
                    onReady={this.setNavigationReady}
```

File: src/libs/actions/User.js [hunk 3]
```diff
@@ -217,28 +217,39 @@ function validateLogin(accountID, validateCode) {
...
-            }
-        } else {
-            const error = lodashGet(response, 'message', 'Unable to validate login.');
-            Onyx.merge(ONYXKEYS.ACCOUNT, {error});
-        }
-    }).finally(() => {
-        Onyx.merge(ONYXKEYS.ACCOUNT, {isLoading: false});
-        Navigation.navigate(redirectRoute);
+    }, {
+        optimisticData: [
```
Comment (flodnv) [inline src/libs/actions/User.js:224]: Are these all the same? Can we make this explicit by defining one of them and then say `successData = optimisticData = failureData`?

**Review by flodnv — COMMENTED** [review]
What's going on with the route things, why did we need these changes here?

**Comment by robertjchen** [comment]
> What's going on with the route things, why did we need these changes here?

Race condition as a result of removing chaining (and the convenient delay that an API call adds). See OP for explanation.

**Comment by robertjchen** [comment]
Updated, and re-tested, thanks!

**Review by NikkiWines — APPROVED** [review]
👍 looks good

**Comment by robertjchen** [comment]
Merging per https://expensify.slack.com/archives/C02HWMSMZEC/p1661369238865979?thread_ts=1661349810.605389&cid=C02HWMSMZEC

Let me know if there's anything else and I can address in a followup PR 👍 #urgency #iterate

**Comment by robertjchen** [comment]
It was?
Screen Shot 2022-08-25 at 1 21 04 AM

**Comment by roryabraham** [comment]
> What's going on with the route things, why did we need these changes here?

@flodnv Also, it has to do with the fact that we have two navigation containers – one for PublicScreens and one for AuthScreens. When one unmounts, we have to wait for the other to be ready before we can navigate to routes again. Better long-term design would be to have one parent navigation container, and conditionally render screen groups instead of conditionally rendering navigation containers.

**Pull request merged**